trees/trees.go

```diff
@@ -0,0 +1,156 @@
+// signers) associated with them.
+package trees
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/rsa"
+	"fmt"
+
+	"github.com/google/trillian"
+	tcrypto "github.com/google/trillian/crypto"
+	"github.com/google/trillian/crypto/keys"
+	"github.com/google/trillian/crypto/sigpb"
+	"github.com/google/trillian/errors"
+	"github.com/google/trillian/merkle"
+	"github.com/google/trillian/merkle/rfc6962"
+	"github.com/google/trillian/storage"
+	"golang.org/x/net/context"
+)
+
+type treeKey struct{}
+
+// NewContext returns a ctx with the given tree.
+func NewContext(ctx context.Context, tree *trillian.Tree) context.Context {
+	return context.WithValue(ctx, treeKey{}, tree)
+}
+
+// FromContext returns the tree within ctx or false.
+func FromContext(ctx context.Context) (*trillian.Tree, bool) {
+	tree, ok := ctx.Value(treeKey{}).(*trillian.Tree)
+	return tree, ok && tree != nil
+}
+
+// GetOpts contains validation options for GetTree.
+type GetOpts struct {
+	// TreeType is the expected type of the tree. Use trillian.TreeType_UNKNOWN_TREE_TYPE to
+	// allow any type.
+	TreeType trillian.TreeType
+	// Readonly is whether the tree will be used for read-only purposes.
+	Readonly bool
+}
+
+// GetTree returns the specified tree, either from the ctx (if present) or read from storage.
+// The tree will be validated according to GetOpts before returned. Tree state is also considered
+// (for example, deleted tree will return NotFound errors).
+func GetTree(ctx context.Context, s storage.AdminStorage, treeID int64, opts GetOpts) (*trillian.Tree, error) {
+	// TODO(codingllama): Record stats of ctx hits/misses, so we can assess whether RPCs work
+	// as intended.
+	tree, ok := FromContext(ctx)
+	if !ok || tree.TreeId != treeID {
+		var err error
+		tree, err = getTreeFromStorage(ctx, s, treeID)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	switch {
+	case opts.TreeType != trillian.TreeType_UNKNOWN_TREE_TYPE && tree.TreeType != opts.TreeType:
+		return nil, errors.Errorf(errors.InvalidArgument, "operation not allowed for %s-type trees", tree.TreeType)
+	case tree.TreeState == trillian.TreeState_FROZEN && !opts.Readonly:
+		return nil, errors.Errorf(errors.FailedPrecondition, "operation not allowed on %s trees", tree.TreeState)
+	case tree.TreeState == trillian.TreeState_SOFT_DELETED || tree.TreeState == trillian.TreeState_HARD_DELETED:
+		return nil, errors.Errorf(errors.NotFound, "unknown tree: %v", tree.TreeId)
+	}
+
+	return tree, nil
+}
+
+func getTreeFromStorage(ctx context.Context, s storage.AdminStorage, treeID int64) (*trillian.Tree, error) {
+	tx, err := s.Snapshot(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer tx.Close()
+	tree, err := tx.GetTree(ctx, treeID)
+	if err != nil {
+		return nil, err
+	}
+	if err := tx.Commit(); err != nil {
+		return nil, err
+	}
+	return tree, nil
+}
+
+// Hash returns the crypto.Hash configured by the tree.
+func Hash(tree *trillian.Tree) (crypto.Hash, error) {
```

Comment:
*Need*, no, but from a completeness PoV I think it's good that it is. Maybe tomorrow someone will want a pure crypto.Hash out of a tree for something?